### PR TITLE
request id header for SwindonChat

### PR DESCRIPTION
**Pull request by popravich**

### CODE REVIEW

File: src/chat/backend.rs [hunk 1]
```diff
@@ -98,19 +112,40 @@ impl CallCodec {
             conn_id: cid,
             destination: destination.clone(),
             sender: sender,
+            request_id_header: request_id_header,
+        }
+    }
+
+    fn add_request_id<S>(&self, e: &mut http::Encoder<S>, cid: &str) {
+        if let Some(ref header) = self.request_id_header {
+            // NOTE: should request_id validation be more strict?
```
Comment (tailhook) [inline src/chat/backend.rs]: Yes. We send it as a header. Not validating it is a security issue.

File: src/chat/backend.rs [hunk 2]
```diff
@@ -98,19 +112,40 @@ impl CallCodec {
...
+            };
+            e.add_header(header, rxid).unwrap();
         }
     }
 }
 
 impl InactivityCodec {
     pub fn new(path: &Arc<String>, sid: &SessionId,
         destination: &Arc<Destination>)
+        // request_id_header: Option<String>)
```
Comment (tailhook) [inline src/chat/backend.rs]: Turn it into a todo, or remove.

Comment (popravich) [inline src/chat/backend.rs]: Hmm, InactivityCodec depends on session pool and not chat handler (settings)
so we either should have another `request-id-header` setting in session-pool, 
or drop it from here.
Maybe we should move `request-id-header` setting from chat & proxy handlers
into http-destination and provide some generic interface to set it for chat/proxy/whatever handlers?

Comment (tailhook) [inline src/chat/backend.rs]: Well, let's move `request-id-header` to destination and deprecate in on handler. Any problems with that?

File: src/chat/backend.rs [hunk 3]
```diff
@@ -105,6 +113,32 @@ impl CallCodec {
             sender: sender,
         }
     }
+
+    fn add_request_id<S>(&self, e: &mut http::Encoder<S>) {
+        if let Some(ref header) = self.destination.request_id_header {
+            let rid = match self.meta.get("request_id") {
+                Some(&Json::String(ref rid)) if rid.len() <= 36 => {
+                    if rid.chars().all(|c| {
```
Comment (tailhook) [inline src/chat/backend.rs]: This check should be a function.

File: src/chat/backend.rs [hunk 4]
```diff
@@ -105,6 +113,32 @@ impl CallCodec {
...
+                    } else {
+                        format!("{}-{}-{}", self.server_id, self.conn_id,
+                            request_id::new())
+                    }
+                }
+                Some(&Json::Number(ref n)) if n.is_u64() => {
+                    format!("{}-{}-{}", self.server_id, self.conn_id,
+                        n.as_u64().unwrap())
+                }
+                _ => {
```
Comment (tailhook) [inline src/chat/backend.rs]: Let's reject invalid request-id with an error and document it. I don't see a good reason to ignore.

Comment (popravich) [inline src/chat/backend.rs]: Reject closing socket?
The only valid `request_id` would be either `u64` or alphanumeric strings up to 36 chars.

Comment (tailhook) [inline src/chat/backend.rs]: It's "Call" not "Auth". Se we might just send error message.

Comment (popravich) [inline src/chat/backend.rs]: Oh, true

**Comment by tailhook** [comment]
Looks good. Still missing docs.

**Pull request merged**